File: String/Day 11-String Basics in C++/string_basics.py
# 1️⃣ Mutable Demonstration

def string_demo():
    print("Python String Demo (Immutable)")

    s = "hello"
    print("Original:", s)

    # Python strings are immutable
    # s[0] = 'H'  ❌ This will cause error

    s = "H" + s[1:]
    print("After modification:", s)
    print()


# 2️⃣ Frequency using list (lowercase only)

def frequency_using_array(s):
    print("Frequency using list (lowercase only)")

    freq = [0] * 26

    for c in s:
        if 'a' <= c <= 'z':
            freq[ord(c) - ord('a')] += 1

    for i in range(26):
        if freq[i] > 0:
            print(chr(i + ord('a')), "->", freq[i])

    print()


# 3️⃣ Frequency using dictionary

def frequency_using_dict(s):
    print("Frequency using dictionary")

    freq = {}

    for c in s:
        freq[c] = freq.get(c, 0) + 1

    for key, value in freq.items():
        print(key, "->", value)

    print()


if __name__ == "__main__":
    string_demo()

    input_string = "abbccc"

    frequency_using_array(input_string)

    frequency_using_dict(input_string)

File: String/Day 11-String Basics in C++/string_basics.cpp
#include <iostream>
#include <string>
#include <vector>
#include <unordered_map>
using namespace std;

// 1️⃣ Demonstrating Character Array
void charArrayDemo() {
    cout << "Character Array Demo\n";

    char name[10] = "Priyal";
    cout << "Original: " << name << endl;

    // Modify character
    name[0] = 'A';
    cout << "After modification: " << name << endl;

    cout << endl;
}

// 2️⃣ Demonstrating string (STL)
void stringDemo() {
    cout << "STL string Demo\n";

    string s = "hello";
    cout << "Original: " << s << endl;

    s[0] = 'H';
    s += " world";

    cout << "After modification: " << s << endl;
    cout << endl;
}

// 3️⃣ Frequency Count using Array (Lowercase only)
void frequencyUsingArray(string s) {
    cout << "Frequency using array (lowercase only)\n";

    vector<int> freq(26, 0);

    for (char c : s) {
        if (c >= 'a' && c <= 'z')
            freq[c - 'a']++;
    }

    for (int i = 0; i < 26; i++) {
        if (freq[i] > 0) {
            cout << char(i + 'a') << " -> " << freq[i] << endl;
        }
    }

    cout << endl;
}

// 4️⃣ Frequency Count using unordered_map
void frequencyUsingMap(string s) {
    cout << "Frequency using unordered_map\n";

    unordered_map<char, int> freq;

    for (char c : s) {
        freq[c]++;
    }

    for (auto &pair : freq) {
        cout << pair.first << " -> " << pair.second << endl;
    }

    cout << endl;
}

int main() {

    charArrayDemo();

    stringDemo();

    string input = "abbccc";

    frequencyUsingArray(input);

    frequencyUsingMap(input);

    return 0;
}
